x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/ml/datafeed/DatafeedUpdate.java

```diff
@@ -0,0 +1,330 @@
+        return scriptFields == null ? Collections.emptyList() : scriptFields;
+    }
+
+    ChunkingConfig getChunkingConfig() {
+        return chunkingConfig;
+    }
+
+    /**
+     * The lists of indices and types are compared for equality but they are not
+     * sorted first so this test could fail simply because the indices and types
+     * lists are in different orders.
+     */
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+
+        if (other == null || getClass() != other.getClass()) {
+            return false;
+        }
+
+        DatafeedUpdate that = (DatafeedUpdate) other;
+
+        return Objects.equals(this.id, that.id)
+            && Objects.equals(this.jobId, that.jobId)
+            && Objects.equals(this.frequency, that.frequency)
+            && Objects.equals(this.queryDelay, that.queryDelay)
+            && Objects.equals(this.indices, that.indices)
+            && Objects.equals(this.types, that.types)
+            && Objects.equals(this.query, that.query)
+            && Objects.equals(this.scrollSize, that.scrollSize)
+            && Objects.equals(this.aggregations, that.aggregations)
+            && Objects.equals(this.scriptFields, that.scriptFields)
+            && Objects.equals(this.chunkingConfig, that.chunkingConfig);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, jobId, frequency, queryDelay, indices, types, query, scrollSize, aggregations, scriptFields,
+            chunkingConfig);
+    }
+
+    boolean isNoop(DatafeedConfig datafeed) {
+        return (frequency == null || Objects.equals(frequency, datafeed.getFrequency()))
+            && (queryDelay == null || Objects.equals(queryDelay, datafeed.getQueryDelay()))
+            && (indices == null || Objects.equals(indices, datafeed.getIndices()))
+            && (types == null || Objects.equals(types, datafeed.getTypes()))
+            && (query == null || Objects.equals(query, datafeed.getQuery()))
+            && (scrollSize == null || Objects.equals(scrollSize, datafeed.getQueryDelay()))
+            && (aggregations == null || Objects.equals(aggregations, datafeed.getAggregations()))
+            && (scriptFields == null || Objects.equals(scriptFields, datafeed.getScriptFields()))
+            && (chunkingConfig == null || Objects.equals(chunkingConfig, datafeed.getChunkingConfig()));
+    }
+
+    public static class Builder {
+
+        private String id;
+        private String jobId;
+        private TimeValue queryDelay;
+        private TimeValue frequency;
+        private List<String> indices;
+        private List<String> types;
+        private QueryBuilder query;
+        private AggregatorFactories.Builder aggregations;
+        private List<SearchSourceBuilder.ScriptField> scriptFields;
+        private Integer scrollSize;
+        private ChunkingConfig chunkingConfig;
+
+        public Builder() {
```

Comment:
This constructor allows a `DatafeedUpdate` with a `null` `id` to be built.  We should prevent this as other code is assuming `id` will not be `null`.  So either remove this constructor, or, if there's a good reason to keep it, add an `Objects.requireNonNull(id)` in the constructor of `DatafeedUpdate` so it's checked when building.